include/pmacc/kernel/warp.hpp

```diff
@@ -53,66 +53,70 @@ namespace pmacc
          * @param data value to broadcast
          * @param srcLaneId lane id of the source thread
-         * @return value send by the source thread
-         *
-         * \{
          */
-        //! broadcast a 32bit integer
-        DINLINE int32_t warpBroadcast(const int32_t data, const int32_t srcLaneId)
-        {
-#        if(__CUDACC_VER_MAJOR__ >= 9)
-            return __shfl_sync(__activemask(), data, srcLaneId);
-#        else
-            return __shfl(data, srcLaneId);
-#        endif
-        }
+        template<typename T_MaskType, typename T_DataType>
+        DINLINE T_DataType warpBroadcast(T_MaskType const mask, T_DataType const data, int const srcLaneId);
 
-        //! Broadcast a 64bit integer by using 2 32bit broadcasts
-        DINLINE int64_cu warpBroadcast(int64_cu data, const int32_t srcLaneId)
-        {
-            int32_t* const pData = reinterpret_cast<int32_t*>(&data);
-            pData[0] = warpBroadcast(pData[0], srcLaneId);
-            pData[1] = warpBroadcast(pData[1], srcLaneId);
-            return data;
-        }
+        /** check if the first type is equal to any other */
+        template<typename T, typename... T_TypeList>
+        inline constexpr bool matchAnyType = (std::is_same_v<T, T_TypeList> || ...);
 
-        //! Broadcast a 32bit unsigned int
-        DINLINE uint32_t warpBroadcast(const uint32_t data, const int32_t srcLaneId)
-        {
-            return static_cast<uint32_t>(warpBroadcast(static_cast<int32_t>(data), srcLaneId));
-        }
 
-        //! Broadcast a 64bit unsigned int
-        DINLINE uint64_cu warpBroadcast(const uint64_cu data, const int32_t srcLaneId)
+        /** fallback for not nativly supported types
+         *
+         * emulates the brotcast by repeating 32bit warp broatcasts
+         */
+        template<typename T_MaskType, typename T_DataType, typename T_Sfinae = void>
+        struct WarpBroatcast
         {
-            return static_cast<uint64_cu>(warpBroadcast(static_cast<int64_cu>(data), srcLaneId));
-        }
+            DINLINE T_DataType operator()(T_MaskType const mask, T_DataType data, int const srcLaneId) const
+            {
+                static_assert(sizeof(T_DataType) >= sizeof(int) && (sizeof(T_DataType) % sizeof(int)) == 0);
+                int* const pData = reinterpret_cast<int*>(&data);
+                for(int i = 0; i < sizeof(T_DataType) / sizeof(int); ++i)
+                    pData[i] = warpBroadcast(mask, pData[i], srcLaneId);
 
-        //! Broadcast a 32bit float
-        DINLINE float warpBroadcast(const float data, const int32_t srcLaneId)
+                return data;
+            }
+        };
+
+        /** CUDA and HIP native supported data types */
+        template<typename T_MaskType, typename T_DataType>
+        struct WarpBroatcast<
+            T_MaskType,
+            T_DataType,
+            std::enable_if_t<matchAnyType<
+                T_DataType,
+                int,
```

Comment:
This would be nice but would make the generator expression more complicated because I need the variable template signature. ONly a workaround with `std::tuple` and unfolding the tuple with `std::Get<>` would come into my mind, but IMO the expression will be much more complicated to read.

Reply:
It isn't too ugly:
```c++
#include <iostream>
#include <type_traits>

template <typename...> struct TypeList {};

using NativeTypes = TypeList<int, float>;

template <typename T, typename... T_TypeList,
          template <typename...> typename T_Pack>
inline constexpr bool matchAnyType(T_Pack<T_TypeList...>) {
  return (std::is_same_v<T, T_TypeList> || ...);
}

template <typename T, typename = void> struct F {
  F() { std::cout << "Called fallback!\n"; }
};

template <typename T>
struct F<T, std::enable_if_t<matchAnyType<T>(NativeTypes{})>> {
  F() { std::cout << "Called implementation!\n"; }
};

int main() {
  std::cout << matchAnyType<int>(NativeTypes{}) << "\n";
  std::cout << matchAnyType<float>(NativeTypes{}) << "\n";
  std::cout << matchAnyType<char>(NativeTypes{}) << "\n";

  F<int>();
  F<float>();
  F<char>();
}
```

Reply:
I will try https://godbolt.org/z/4nzG1eree based on your solution

Reply:
Nice one! Didn't know about `std::disjunction`!